cli/tools/script.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use crate::colors;
+use crate::config_file::ConfigFile;
+use crate::flags::Flags;
+use crate::flags::ScriptFlags;
+use crate::proc_state::ProcState;
+use deno_core::anyhow::bail;
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use std::collections::HashMap;
+use std::process::Command;
+use std::sync::Arc;
+
+fn get_scripts_config(
+  maybe_config_file: Option<&ConfigFile>,
+) -> Result<HashMap<String, String>, AnyError> {
+  if let Some(config_file) = maybe_config_file {
+    let maybe_scripts_config = config_file.to_scripts_config()?;
+    if let Some(scripts_config) = maybe_scripts_config {
+      Ok(scripts_config)
+    } else {
+      bail!("No configured scripts")
+    }
+  } else {
+    bail!("No config file found")
+  }
+}
+
+fn print_available_scripts(scripts_config: HashMap<String, String>) {
+  eprintln!("{}", colors::green("Available scripts:"));
+
+  let mut script_names: Vec<String> =
+    scripts_config.clone().into_keys().collect();
+  script_names.sort();
+
+  for name in script_names {
+    eprintln!("- {}", colors::cyan(&name));
+    eprintln!("    {}", scripts_config[&name])
+  }
+}
+
+pub async fn execute_script(
+  flags: Flags,
+  script_flags: ScriptFlags,
+) -> Result<i32, AnyError> {
+  let flags = Arc::new(flags);
+  let ps = ProcState::build(flags.clone()).await?;
+  let scripts_config = get_scripts_config(ps.maybe_config_file.as_ref())?;
+  let config_file_url = &ps.maybe_config_file.as_ref().unwrap().specifier;
+  let config_file_path = if config_file_url.scheme() == "file" {
+    config_file_url.to_file_path().unwrap()
+  } else {
+    bail!("Only local configuration files are supported")
+  };
+
+  if script_flags.script.is_empty() {
+    print_available_scripts(scripts_config);
+    return Ok(1);
+  }
+
+  let cwd = config_file_path.parent().unwrap();
+  let script_name = script_flags.script;
+  let maybe_script = scripts_config.get(&script_name);
+
+  if let Some(script) = maybe_script {
+    let (shell_name, shell_flag) = if cfg!(windows) {
+      // TODO: allow to use PowerShell?
+      ("cmd", "/C")
+    } else {
+      // TODO: allow to use other shells?
+      ("sh", "-c")
+    };
+    let shell_arg = format!("{} {}", script, flags.argv.join(" "));
```

Comment:
I'm not sure we should straight pass the script to the shell. A really annoying thing about cmd and powershell is that setting the environment variables are annoying. In npm, this is typically solved by installing the cross-env package: https://www.npmjs.com/package/cross-env -- I think whatever solution we come up with should solve this.

Reply:
That's a good point and something that @lucacasonato raised to me as well. I'll look into that.